src/snowflake/snowpark/session.py

```diff
@@ -710,17 +710,23 @@ def _resolve_packages(
         )
         for package in packages:
             if isinstance(package, ModuleType):
-                package_name = MODULE_NAME_TO_PACKAGE_NAME_MAP.get(
+                package = MODULE_NAME_TO_PACKAGE_NAME_MAP.get(
                     package.__name__, package.__name__
                 )
-                package = f"{package_name}=={pkg_resources.get_distribution(package_name).version}"
+                package = (
+                    f"{package}=={pkg_resources.get_distribution(package).version}"
+                )
                 use_local_version = True
             else:
                 package = package.strip().lower()
                 use_local_version = False
             package_req = pkg_resources.Requirement.parse(package)
-            # get the standard package name
-            package_name = package_req.key
+            # get the standard package name if there is no underscore
+            # underscores are discouraged in package names, but are still used in Anaconda channel
+            # pkg_resources.Requirement.parse will convert all underscores to dashes
+            package_name = (
+                package if not use_local_version and "_" in package else package_req.key
```

Comment:
We only use `package_req` when the `if` check fails, should we only parse in the `else` branch?

e.g. `package_name = package if ... else pkg_resources.Requirement.parse(package).key`, or we could make an explicit `if/else` if this is not very readable.

Reply:
It's also used when validating packages https://github.com/snowflakedb/snowpark-python/blob/b8bc6b193b21fd8cb2d779034057ad3308ae74c3/src/snowflake/snowpark/session.py#L751

Reply:
Ah sorry missed that lol.